### ISSUE TITLE
feat(agents): optimize token consumption across agent definitions, skills, and context injection

### ISSUE THREAD

**Issue by cmtzco** [co-labs-co/context-harness#99]
## Summary

Research into prompt compression techniques reveals that ContextHarness agent definitions carry **~6x more tokens than necessary** for the same behavioral outcomes. The `.claude/` template variants already prove this — encoding identical agent behaviors in 6-22x fewer characters than their `.opencode/` counterparts. This issue proposes a structured optimization effort across all token-consuming surfaces.

## Problem

Every ContextHarness conversation has a **baseline token overhead** before the user says a single word. As the framework grows (more skills, richer agent definitions, longer AGENTS.md), this overhead compounds — eating into the user's context window and increasing API costs.

### Current Token Budget (measured with `cl100k_base` tokenizer)

| Scenario | Framework Overhead | Notes |
|----------|-------------------:|-------|
| Normal conversation | ~5,137 tokens | AGENTS.md + primary agent |
| With 2-3 skills loaded | ~11,800 tokens | + on-demand skills |
| `/compact` cycle | ~9,800 tokens | + compaction guide agent |
| `/baseline` orchestrator | ~37,200 tokens | Largest single-conversation load |
| `/baseline` all workers | ~150,000+ tokens | 8 agents × parallel workers |

### Token Consumption by Category

| Category | Files | Total ~Tokens | Loading Behavior |
|----------|------:|-------------------:|------------------|
| AGENTS.md (root) | 1 | **3,727** | Every conversation |
| Agent definitions | 12 | **37,268** | Mixed |
| Command definitions | 7 | **10,263** | On-demand per command |
| Skill definitions | 15 | **45,194** | On-demand, lazy-loaded |
| Session files | 25 avg | **~900 avg** | One per active session |

### The `.opencode/` vs `.claude/` Disparity

We already ship both formats in our templates. The size difference is stark:

| Agent | `.opencode/` chars | `.claude/` chars | Ratio |
|-------|:-----------:|:----------:|:-----:|
| compaction-guide | 13,424 | 1,959 | **6.9x** |
| baseline.md command | 27,597 | 1,271 | **21.7x** |
| Agent definitions (avg) | ~12,400 | ~2,000 | **~6x** |

This proves the semantic content can be expressed in dramatically fewer tokens. The `.opencode/` format is verbose — it repeats patterns, includes extensive examples, and uses full prose where structured shorthand would suffice.

## Research Context

We evaluated several compression approaches:

| Approach | Compression | API-Compatible? | Dependencies | Verdict |
|----------|:-----------:|:---------------:|:------------:|---------|
| **LLMLingua (neural)** | 5-20x | ✅ | PyTorch (~2.8 GB) | ❌ Too heavy for CLI tool |
| **Gist tokens (soft prompts)** | 26x | ❌ Needs fine-tuning | Model access | ❌ Not applicable |
| **Hieratic-style shorthand** | 50-70% | ✅ | None | ✅ Worth investigating |
| **Agent definition rewrite** | 5-10x | ✅ | None | ✅ **Best ROI** |
| **Extractive pruning** | 2-5x | ✅ | None | ✅ For SESSION.md |
| **tiktoken audit tooling** | N/A | N/A | ~3 MB | ✅ For measurement |

**Industry precedent**: No major AI coding tool (Cursor, Claude Code, OpenCode) uses neural prompt compression. They all use **context engineering** — selective inclusion, lazy loading, and concise authoring. Anthropic reported Claude Code achieves **5.5x fewer tokens** than Cursor through better context selection alone.

### Phase 1: Measure (prerequisite for everything)

**Add a `context-harness audit` CLI command** that scans a project's agent definitions, skills, commands, AGENTS.md, and session files, then reports token counts using `tiktoken`.

```
$ context-harness audit

📊 Token Budget Report
─────────────────────────────────────────────
Component                     Tokens   Load
─────────────────────────────────────────────
AGENTS.md                      3,727   always
context-harness.md (agent)     1,410   always
─────────────────────────────────────────────
Always-on overhead             5,137

Skills (15 installed)         45,194   lazy
  python-result-pattern        2,728
  python-cli-click             2,644
  ...

Commands (7)                  10,263   on-demand
  baseline.md                  6,899   ⚠️ largest
  ctx.md                         779
  ...

Sessions (25)                    900   one active
─────────────────────────────────────────────
Worst case (baseline run)    ~37,200
─────────────────────────────────────────────
```

Implementation: Add `tiktoken` as an optional dependency. Use character-based estimation (`len/4`) as fallback.

### Phase 2: Compress Agent Definitions

Systematically rewrite `.opencode/` agent definitions to be more concise while preserving behavior. Target: **3x reduction** (from ~12,400 to ~4,000 chars average).

Techniques:
- Replace verbose prose with structured rules/tables
- Remove redundant examples (keep 1, not 3)
- Extract shared patterns into a base template referenced by agents
- Use consistent shorthand conventions documented once in AGENTS.md
- Remove ASCII art diagrams where a compact description suffices (e.g., `baseline.md` has 50+ lines of box-drawing characters)

**Acceptance criteria**: Agent behavior is unchanged when tested against the same scenarios. Measured by token count before/after.

### Phase 3: Optimize AGENTS.md

AGENTS.md is loaded **every single conversation** — it's the highest-frequency token consumer. Target: **40% reduction** (~3,727 → ~2,200 tokens).

- Trim the "Available Skills" section to just trigger keywords + names (remove descriptions already in SKILL.md)
- Remove the full directory tree (agents can discover structure on demand)  
- Condense code standard tables
- Remove "Quick Reference" section that duplicates info available in help commands

### Phase 4: Investigate Compact Session Format

Evaluate a structured shorthand format for SESSION.md output from `/compact`. Instead of verbose prose:

```markdown
## Current State
The implementation of the skill service is complete. We have added all the 
necessary methods including install, upgrade, and list operations. The tests 
are passing and cover success, failure, and edge cases.
```

Use a compact structured format:

```markdown
## State
- skill_service: complete (install, upgrade, list)
- tests: passing (success/failure/edge cases covered)
```

### Phase 5: Shared Base Templates for Agents

Many agent definitions repeat the same boilerplate (tool access patterns, output formatting, error handling). Extract shared patterns into a base template that agents reference, eliminating per-agent duplication.

Current: Each of the 8 baseline agents repeats ~3,000 tokens of shared context.
Target: Shared base (~1,500 tokens) + agent-specific delta (~1,500 tokens) = ~3,000 per agent vs ~6,000.

## Success Metrics

| Metric | Current | Target | Method |
|--------|--------:|-------:|--------|
| Always-on overhead | 5,137 tokens | < 3,000 tokens | `context-harness audit` |
| Avg agent definition | ~9,300 tokens | < 3,500 tokens | Token count |
| `/baseline` orchestrator | ~37,200 tokens | < 18,000 tokens | Token count |
| `/baseline` total workers | ~150,000 tokens | < 80,000 tokens | Token count |
| AGENTS.md | 3,727 tokens | < 2,200 tokens | Token count |

## Non-Goals

- **Neural prompt compression** (LLMLingua, etc.) — Too heavy for a CLI tool (+2.8 GB deps, 10+ second latency)
- **Soft prompt / gist token approaches** — Requires model fine-tuning, not API-compatible
- **Changing the lazy-loading architecture** — It's already well-designed; this optimizes what gets loaded, not when

## Related

- #83 — Context overflow during baseline (symptom of this problem)
- #82 — Baseline subagent invocation patterns (affected by token budgets)

## References

- [LLMLingua-2: Microsoft Research](https://arxiv.org/abs/2403.12968) — Neural prompt compression (evaluated, rejected for weight)
- [Base64Bench: LessWrong](https://www.lesswrong.com/posts/5F6ncBfjh2Bxnm6CJ) — Encoding schemes and LLM comprehension
- [Fundamental Limits of Prompt Compression: NeurIPS 2024](https://proceedings.neurips.cc/paper_files/paper/2024/file/ac8fbba029dadca99d6b8c3f913d3ed6-Paper-Conference.pdf) — Rate-distortion framework
- [Stingy Context: 18:1 Code Compression](https://www.researchgate.net/publication/400177311) — Hierarchical compression for code contexts
- [Promptfoo Base64 Strategy](https://www.promptfoo.dev/docs/red-team/strategies/base64/) — Encoding vs. safety training gap
